lib/models/package-info-cache.js

```diff
@@ -0,0 +1,1049 @@
+  }
+
+  /**
+   * Add a reference to an internal addon PackageInfo object.
+   * "internal" addons (note: not in-repo addons) only exist in
+   * Projects, not other packages. Since the cache is loaded from
+   * 'loadProject', this can be done appropriately.
+   *
+   * @private
+   * @method _addInternalAddon
+   * @param {PackageInfo} internalAddonPkg the PackageInfo for the internal addon
+   * @return null
+   */
+  _addInternalAddon(internalAddonPkg) {
+    if (!this.internalAddons) {
+      this.internalAddons = [];
+    }
+    this.internalAddons.push(internalAddonPkg);
+  }
+
+  /**
+   * For each dependency in the given list, find the corresponding
+   * PackageInfo object in the cache (going up the file tree if
+   * necessary, as in the node resolution algorithm). Return a map
+   * of the dependencyName to PackageInfo object. Caller can then
+   * store it wherever they like.
+   *
+   * Note: this is not to be  called until all packages that can be have
+   * been added to the cache.
+   *
+   * Note: this is for ALL dependencies, not just addons. To get just
+   * addons, filter the result by calling pkgInfo.isAddon().
+   *
+   * @private
+   * @method _addDependencies
+   * @param {PackageInfo} dependencies value of 'dependencies' or 'devDependencies'
+   *        attributes of a package.json.
+   * @return {Object} a JavaScript object keyed on dependency name/path with
+   *    values the corresponding PackageInfo object from the cache.
+   */
+  _addDependencies(dependencies) {
+    if (!dependencies) {
+      return null;
+    }
+
+    let dependencyNames = Object.keys(dependencies);
+
+    if (dependencyNames.length === 0) {
+      return null;
+    }
+
+    let packages = Object.create(null);
+
+    let missingDependencies = [];
+
+    dependencyNames.forEach(dependencyName => {
+      let dependencyPackage;
+
+      // much of the time the package will have dependencies in
+      // a node_modules inside it, so check there first because it's
+      // quicker since we have the reference. Only check externally
+      // if we don't find it there.
+      if (this.nodeModules) {
+        dependencyPackage = this.nodeModules._findPackage(dependencyName);
+      }
+
+      if (!dependencyPackage) {
+        dependencyPackage = this.cache._findPackage(
+          dependencyName,
+          path.dirname(this.realPath)
+        );
+      }
+
+      if (dependencyPackage) {
+        packages[dependencyName] = dependencyPackage;
+      } else {
+        missingDependencies.push(dependencyName);
+      }
+    });
+
+    if (missingDependencies.length > 0) {
+      this._addError(ERROR_DEPENDENCIES_MISSING, missingDependencies);
+    }
+
+    return packages;
+  }
+
+  isAddon() {
+    let val =
+      Array.isArray(this.pkg.keywords) &&
+      this.pkg.keywords.indexOf('ember-addon') >= 0;
+    return val;
+  }
+
+  // XXX Should this move to Addon.js (and change appropriately to use the cache here?)
```

Comment:
should this comment be addressed before merging?

Reply:
Probably. At this point I'm going to leave it, so will consider just removing the function entirely, since things are working and addon-factory already does what this is doing (since I more or less copied from there). We could revisit later on a new effort.

Reply:
Removed the function, and the reference to 'this.addonConstructor' in the members of PackageInfo.